metashare/api/views.py

```diff
@@ -1,19 +1,53 @@
 from django.contrib.auth import get_user_model
-from rest_framework import generics
+from rest_framework import generics, status, viewsets
 from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.views import APIView
 
-from .serializers import FullUserSerializer
+from . import gh
+from .models import GitHubRepository, Product
+from .paginators import ProductPaginator
+from .serializers import FullUserSerializer, ProductSerializer
 
 User = get_user_model()
 
 
-class UserView(generics.RetrieveAPIView):
-    model = User
-    serializer_class = FullUserSerializer
-    permission_classes = (IsAuthenticated,)
-
+class CurrentUserObjectMixin:
     def get_queryset(self):
         return self.model.objects.filter(id=self.request.user.id)
 
     def get_object(self):
         return self.get_queryset().get()
+
+
+class UserView(CurrentUserObjectMixin, generics.RetrieveAPIView):
+    model = User
+    serializer_class = FullUserSerializer
+    permission_classes = (IsAuthenticated,)
+
+
+class UserRefreshView(CurrentUserObjectMixin, APIView):
+    model = User
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request):
+        user = self.get_object()
+        repos = gh.get_all_org_repos(user)
+        GitHubRepository.objects.filter(user=user).delete()
+        GitHubRepository.objects.bulk_create(
+            [GitHubRepository(user=user, url=repo) for repo in repos]
+        )
+        serializer = FullUserSerializer(user)
+        # TODO: this _should_ return the new list of products for the user, but that
+        # feels wrong for this endpoint. Merits thought.
+        return Response(serializer.data, status=status.HTTP_200_OK)
```

Comment:
@wlonk I'm curious how you anticipate this view being used.

Reply:
I don't really know; I just felt uneasy having no way to refresh what products you can see without logging in and out again.

Reply:
Hm. So it won't be refreshed when the `products` api is requested? We're only doing that once per page-load, so I would expect it to be up-to-date...

Reply:
That's what I was saying about not wanting HTTP requests inside our HTTP requests; I compromised (with myself) by having a dedicated "this could be slow" endpoint to refresh thing.

Reply:
Hm, that doesn't seem useful from a user's perspective. I think this data should be up-to-date on a page-refresh, not just a login/logout. So how does it help to require an additional API call to refresh the data first?

Reply:
I just really really don't want to query GH each time `.../products/` is hit. It seems bad. Maybe it's not so bad, maybe there's another way. I don't know yet.

Reply:
Ok, I'm open to that. Maybe we want to add an explicit "Refresh repos" btn, similar to Pyup or Greenkeeper or Coveralls.io.

Reply:
@stacyk do you wanna weigh in on the UX considerations there?

To recap:

We can't (we don't think) reasonably keep the API constantly in sync with GitHub's view of the world, so we have to cache data. Sometimes that cache may be stale, and we'd like to offer users a way to refresh it manually.

Reply:
I think refreshing when the user logs in would be a reasonable compromise, because then there's at least a simple action we can tell users to take if they know it's not up to date.

Reply:
For comparison with how some other tools handle this. If we have the endpoint, I don't see a good reason not to add a button rather than instructions to logout/login.

![Screenshot 2019-06-27 15 49 43](https://user-images.githubusercontent.com/552316/60296200-95403a00-98f3-11e9-8b45-4c63a9134a25.png)

---

![Screenshot 2019-06-27 15 50 27](https://user-images.githubusercontent.com/552316/60296201-95403a00-98f3-11e9-80ff-20cbf1ffb8e6.png)

---

![Screenshot 2019-06-27 15 50 50](https://user-images.githubusercontent.com/552316/60296202-95403a00-98f3-11e9-8400-6485241f50c0.png)

Reply:
I agree, a button sounds good.

Reply:
I wonder how often products would be hit compared to the other pages. My guess is people would spend the majority of their time on a project detail page. With that in mind, the negative of contacting GH each time the product list page is loaded seems like a smaller UX cost than having the user hit (hopefully hit) a button when they want to see any updates? 
It sounds like the consensus is to add a button at this point though.

Reply:
> I wonder how often products would be hit compared to the other pages. My guess is people would spend the majority of their time on a project detail page. With that in mind, the negative of contacting GH each time the product list page is loaded seems like a smaller UX cost than having the user hit (hopefully hit) a button when they want to see any updates?
> It sounds like the consensus is to add a button at this point though.

@stacyk I think either way we won't hit the endpoint every time the products-list page is loaded, but the question is whether we hit it on every page-refresh (typically once/session) or just on login (which may last through multiple sessions).

Reply:
@jgerigmeyer  To clarify, is this: 
> but the question is whether we hit it on every page-refresh (typically once/session)

not the same as:

> hit the endpoint every time the products-list page is loaded

Reply:
@stacyk Ah, well it could be or could not be, depending what you meant by "loaded". 😄 We certainly won't re-fetch during typical navigation through the application (like navigating to a product and then back to the products-list) -- which doesn't reload the browser. So it would only be on initial page-load or after a browser refresh/reload, vs only on login.